diff -Naur v4l-dvb-saa716x-ac1783f4bd1b.orig/linux/drivers/media/common/saa716x/Makefile v4l-dvb-saa716x-ac1783f4bd1b/linux/drivers/media/common/saa716x/Makefile
--- v4l-dvb-saa716x-ac1783f4bd1b.orig/linux/drivers/media/common/saa716x/Makefile	2012-01-29 20:17:10.000000000 +0100
+++ v4l-dvb-saa716x-ac1783f4bd1b/linux/drivers/media/common/saa716x/Makefile	2012-03-03 22:43:45.675807125 +0100
@@ -24,3 +24,4 @@
 obj-$(CONFIG_DVB_SAA716X_FF)	  += saa716x_ff.o
 
 EXTRA_CFLAGS = -Idrivers/media/dvb/dvb-core/
+EXTRA_CFLAGS += -Idrivers/media/dvb/frontends/
diff -Naur v4l-dvb-saa716x-ac1783f4bd1b.orig/linux/drivers/media/common/saa716x/saa716x_ff_ir.c v4l-dvb-saa716x-ac1783f4bd1b/linux/drivers/media/common/saa716x/saa716x_ff_ir.c
--- v4l-dvb-saa716x-ac1783f4bd1b.orig/linux/drivers/media/common/saa716x/saa716x_ff_ir.c	2012-01-29 20:17:10.000000000 +0100
+++ v4l-dvb-saa716x-ac1783f4bd1b/linux/drivers/media/common/saa716x/saa716x_ff_ir.c	2012-03-03 22:44:05.099806606 +0100
@@ -1,3 +1,4 @@
+#include <linux/version.h>
 /*
  * Driver for the remote control of the TT6400 DVB-S2 card
  *
diff -Naur v4l-dvb-saa716x-ac1783f4bd1b.orig/linux/drivers/media/common/saa716x/saa716x_ff_main.c v4l-dvb-saa716x-ac1783f4bd1b/linux/drivers/media/common/saa716x/saa716x_ff_main.c
--- v4l-dvb-saa716x-ac1783f4bd1b.orig/linux/drivers/media/common/saa716x/saa716x_ff_main.c	2012-01-29 20:17:10.000000000 +0100
+++ v4l-dvb-saa716x-ac1783f4bd1b/linux/drivers/media/common/saa716x/saa716x_ff_main.c	2012-03-03 22:44:05.099806606 +0100
@@ -1,3 +1,4 @@
+#include <linux/version.h>
 #include <linux/module.h>
 #include <linux/moduleparam.h>
 #include <linux/kernel.h>
