diff -bur vdr-1.5.6_orig/tools.h vdr-1.5.6/tools.h
--- vdr-1.5.6_orig/tools.h	2007-07-22 18:54:14.000000000 +0200
+++ vdr-1.5.6/tools.h	2007-07-22 18:56:30.000000000 +0200
@@ -175,9 +175,9 @@
 char *strreplace(char *s, const char *s1, const char *s2); ///< re-allocates 's' and deletes the original string if necessary!
 inline char *skipspace(const char *s)
 {
-  if (*s > ' ') // most strings don't have any leading space, so handle this case as fast as possible
+  if ((uchar)*s > ' ') // most strings don't have any leading space, so handle this case as fast as possible
      return (char *)s;
-  while (*s && *s <= ' ') // avoiding isspace() here, because it is much slower
+  while (*s && (uchar)*s <= ' ') // avoiding isspace() here, because it is much slower
         s++;
   return (char *)s;
 }

