diff -Naur vdr-plugin-graphtftng-b38c057267c53b93137a422c14d86236788dc121.orig/graphtftng.c vdr-plugin-graphtftng-b38c057267c53b93137a422c14d86236788dc121/graphtftng.c
--- vdr-plugin-graphtftng-b38c057267c53b93137a422c14d86236788dc121.orig/graphtftng.c	2014-02-21 17:53:24.353685620 +0100
+++ vdr-plugin-graphtftng-b38c057267c53b93137a422c14d86236788dc121/graphtftng.c	2014-02-21 17:55:45.940361241 +0100
@@ -19,6 +19,8 @@
 #include <graphtftng.h>
 #include <span.h>
 
+#define DATA_DIR "/usr/share/vdr/graphTFT"
+
 //***************************************************************************
 // cGraphTFTMenu
 //***************************************************************************
@@ -250,8 +252,8 @@
    logDevice = GraphTFTSetup.LogDevice;
    logLevel = GraphTFTSetup.Level;
 
-   GraphTFTSetup.themesPath = ConfigDirectory() + string("/graphtftng/themes/");
-   GraphTFTSetup.configPath = ConfigDirectory() + string("/graphtftng/");
+   GraphTFTSetup.themesPath = strdup(DATA_DIR) + string("/themes/");
+   GraphTFTSetup.configPath = strdup(DATA_DIR);
 
    // load the thems
 
diff -Naur vdr-plugin-graphtftng-b38c057267c53b93137a422c14d86236788dc121.orig/imlibrenderer/imlibrenderer.c vdr-plugin-graphtftng-b38c057267c53b93137a422c14d86236788dc121/imlibrenderer/imlibrenderer.c
--- vdr-plugin-graphtftng-b38c057267c53b93137a422c14d86236788dc121.orig/imlibrenderer/imlibrenderer.c	2014-02-21 17:53:24.347018953 +0100
+++ vdr-plugin-graphtftng-b38c057267c53b93137a422c14d86236788dc121/imlibrenderer/imlibrenderer.c	2014-02-21 18:02:33.583720355 +0100
@@ -186,7 +186,7 @@
    
    // at least add the default path
 
-   asprintf(&path, "%s/graphtftng/fonts/", confPath.c_str());
+   asprintf(&path, "%s/fonts/", confPath.c_str());
    tell(0, "Info: Adding font path '%s'", path);
    imlib_add_path_to_font_path(path);
    free(path);
