diff -Naur dvbhddevice_0.0.4dag1/dvbhddevice.c dvbhddevice_0.0.4dag2/dvbhddevice.c
--- dvbhddevice_0.0.4dag1/dvbhddevice.c	2011-04-30 21:17:28.000000000 +0200
+++ dvbhddevice_0.0.4dag2/dvbhddevice.c	2011-05-03 22:35:06.000000000 +0200
@@ -10,7 +10,7 @@
 #include "dvbhdffdevice.h"
 #include "setup.h"
 
-static const char *VERSION        = "0.0.4dag1";
+static const char *VERSION        = "0.0.4dag2";
 static const char *DESCRIPTION    = "HD Full Featured DVB device";
 
 #define MAXDVBDEVICES  8
diff -Naur dvbhddevice_0.0.4dag1/dvbhdffdevice.c dvbhddevice_0.0.4dag2/dvbhdffdevice.c
--- dvbhddevice_0.0.4dag1/dvbhdffdevice.c	2011-04-30 12:25:45.000000000 +0200
+++ dvbhddevice_0.0.4dag2/dvbhdffdevice.c	2011-05-03 22:32:00.000000000 +0200
@@ -27,7 +27,6 @@
 int cDvbHdFfDevice::devHdffOffset = -1;
 
 cDvbHdFfDevice::cDvbHdFfDevice(int Adapter, int Frontend)
-:cDevice()
 {
   spuDecoder = NULL;
   audioChannel = 0;
@@ -46,7 +45,7 @@
 
   isHdffPrimary = false;
   if (devHdffOffset < 0) {
-     devHdffOffset = adapter;
+     devHdffOffset = DeviceNumber();
      isHdffPrimary = true;
      mHdffCmdIf = new HDFF::cHdffCmdIf(fd_osd);
      mHdffCmdIf->CmdAvSetAudioDelay(gHdffSetup.AudioDelay);
@@ -108,6 +107,7 @@
 {
   if (!spuDecoder && IsPrimaryDevice())
      spuDecoder = new cDvbSpuDecoder();
+
   return spuDecoder;
 }
 
@@ -737,5 +737,6 @@
         return device->mHdffCmdIf;
      }
   return NULL;
+
 }
 
