--- Makefile-org	2009-03-16 23:46:28.000000000 +0100
+++ Makefile	2009-03-16 23:48:14.000000000 +0100
@@ -10,7 +10,7 @@
 PLUGIN = reelbox
 
 # set it if you want to compile the skin for use with the reelbox
-REELSKIN=1
+#REELSKIN=1
 
 ### The object files (add further files here):
 
@@ -33,13 +33,13 @@
 LIBDIR = ../../lib
 TMPDIR = /tmp
 
-BSPSHM = ../../../utils/bspshm
-HDSHM = ../../../utils/hdshm3/src
+BSPSHM = ../bspshm
+HDSHM = ../hdshm3/src
 
-BSPINCLUDE = -I$(BSPSHM) -I$(BSPSHM)/include
-HDINCLUDE = -I$(HDSHM) -I$(HDSHM)/include
-LIBMAD     ?= ../../../../temp/docimage/libs/libmad
-LIBASOUND  ?= ../../../../temp/docimage/libs/alsa-lib
+#BSPINCLUDE = -I$(BSPSHM) -I$(BSPSHM)/include
+#HDINCLUDE = -I$(HDSHM) -I$(HDSHM)/include
+#LIBMAD     ?= ../../../../temp/docimage/libs/libmad
+#LIBASOUND  ?= ../../../../temp/docimage/libs/alsa-lib
 
 ### Allow user defined options to overwrite defaults:
 
@@ -47,8 +47,9 @@
 
 ### Includes and Defines (add further entries here):
 
-INCLUDES += -I$(LIBASOUND)/include $(BSPINCLUDE) $(HDINCLUDE)
-INCLUDES += -I$(LIBMAD) `freetype-config --cflags`
+#INCLUDES += -I$(LIBASOUND)/include $(BSPINCLUDE) $(HDINCLUDE)
+#INCLUDES += -I$(LIBMAD) `freetype-config --cflags`
+INCLUDES += -I$(BSPSHM) -I$(HDSHM) `freetype-config --cflags`
 
 ifdef REELSKIN
   DEFINES += -DREELSKIN
