diff -Naur vdr-reelbox-3.orig/VideoPlayerPipHd.c vdr-reelbox-3/VideoPlayerPipHd.c
--- vdr-reelbox-3.orig/VideoPlayerPipHd.c	2012-04-09 08:23:28.000000000 +0200
+++ vdr-reelbox-3/VideoPlayerPipHd.c	2012-04-09 08:25:37.000000000 +0200
@@ -492,7 +492,11 @@
 	{
 		int len,gotPicture;
 
-		len = avcodec_decode_video(av_context, decoded_frame, &gotPicture, (uint8_t*)esbuf+esdec, eslen-esdec);	    
+		AVPacket pkt;
+		av_init_packet(&pkt);
+		pkt.data = (uint8_t*)esbuf+esdec;
+		pkt.size = eslen-esdec;
+		len = avcodec_decode_video2(av_context, decoded_frame, &gotPicture, &pkt);
 
 		if (len>0)
 			esdec+=len;
