diff -Naur vdr-1.7.17.orig/Make.config.template vdr-1.7.17/Make.config.template
--- vdr-1.7.17.orig/Make.config.template	2011-03-26 11:44:58.000000000 +0100
+++ vdr-1.7.17/Make.config.template	2011-03-26 11:45:56.000000000 +0100
@@ -73,6 +73,7 @@
 #VALIDINPUT = 1
 #WAREAGLEICON = 1
 #YAEPG = 1
+#JUMPPLAY = 1
 ### You don't need to touch the following:
 
 ifdef DVBDIR
@@ -190,3 +191,7 @@
 ifdef YAEPG
 DEFINES += -DUSE_YAEPG
 endif
+
+ifdef JUMPPLAY
+DEFINES += -DUSE_JUMPPLAY
+endif
\ No newline at end of file
